<img src="https://i.postimg.cc/ZRSDQWPG/pokeapi-pyspark.png" width = "500"/> <center>

# Consumindo e analisando dados da PokéApi utilizando Pyspark

## Importação de bibliotecas

In [0]:
import requests
import json
from pyspark.sql import SparkSession
from pyspark.sql.functions import col, coalesce, count, explode_outer, when, row_number, sum as _sum, min as _min, max as _max, round as _round
from pyspark.sql.window import Window

#Create PySpark SparkSession
spark = SparkSession.builder.appName("Spark App").getOrCreate()

## Extração dos dados da API

Para fazer a extração dos dados, o primeiro passo foi interpretar as informações contidas na [Documentação](https://pokeapi.co/docs/v2) da PokéAPI, os recursos nela disponíveis e seus atributos, ao passo que foram levantadas as informações necessárias para responder às perguntas propostas pelo problema.
Notou-se que todos os dados necessários estavam contidos em alguns Endpoints distribuídos nos seguintes grupos:
- [Evolution](https://pokeapi.co/docs/v2#evolution-section)
- [Pokémon](https://pokeapi.co/docs/v2#pokemon-section)
- [Games](https://pokeapi.co/docs/v2#games-section)

Foi, então, criada uma função para automatizar o processo de fazer as requisições para cada recurso disponível nos Endpoints desejados até que todos os recursos fossem extraídos. Para isso, a função obtém primeiramente o número de recursos disponibilizados pela API em cada Endpoint e utiliza esse valor para obter as URLs de cada recurso e fazer as requisições de dados para cada um deles. Em seguida, é gerado um Spark Dataframe com os dados extraídos.

In [0]:
# Criando função geral de extração de dados e geração de dataframes
def generate_dataframe_from_request(endpoint_name):
    """
    Função que faz a requisição na PokéAPI de todos os
    recursos de um endpoint recebido como parâmetro
    (endpoit_name) e gera um spark dataframe com os dados.
    """
    
    # Faz a primeira requisição para obter o número de recursos disponíveis
    max_page = (requests.get(f'https://pokeapi.co/api/v2/{endpoint_name}').json())["count"]

    # Obtém as urls de todos os recursos disponíveis no Endpoint
    response = requests.get(f'https://pokeapi.co/api/v2/{endpoint_name}?limit={max_page}')
    data = response.json()

    responses_list = []

    # Faz a requisição na API dos dados de cada recurso por meio das urls
    # e compõe uma lista com as respostas de todas as requisições
    for resource in data['results']:
        response = requests.get(resource["url"]).json()
        response_str_json = json.dumps(response)
        responses_list.extend([response_str_json])

    # Gerando raw dataframe
    responses_list_json = sc.parallelize(responses_list)
    df_raw = spark.read.json(responses_list_json, multiLine = True)
    return df_raw

Os Endpoints necessários para responder às perguntas são:
- [Pokémon](https://pokeapi.co/docs/v2#pokemon)
- [Pokémon-Species](https://pokeapi.co/docs/v2#pokemon-species)
- [Evolution-Chain](https://pokeapi.co/docs/v2#evolution-chains)
- [Types](https://pokeapi.co/docs/v2#types)
- [Version-Groups](https://pokeapi.co/docs/v2#version-groups)

In [0]:
# Executando função criada para cada Endpoint que vamos precisar
df_pokemon_raw = generate_dataframe_from_request("pokemon")
df_pokemon_species_raw = generate_dataframe_from_request("pokemon-species")
df_evol_chains_raw = generate_dataframe_from_request("evolution-chain")
df_types_raw = generate_dataframe_from_request("type")
df_version_groups_raw = generate_dataframe_from_request("version-group")

## Modelagem dos dados

### Endpoint "Pokémon"

#### Pokémon
- Dataframe que descreve as principais características de um Pokémon.
- Possui também informações de `species_id` e `evol_chain_id`, que relacionam esse dataframe com o `df_evol_chains_species`.
- *Cada "Pokémon" possui apenas um `species_id` e um `evol_chain_id`.*

OBS: Foi observado que, para todos os Pokémon em `df_pokemon` que estão em sua forma padrão, ou seja,\
`is_default = True`, temos que `pokemon_id = species_id`.

In [0]:
df_pokemon = df_pokemon_raw.selectExpr('id as pokemon_id',
                                       'name as pokemon_name',
                                       'is_default',
                                       'height',
                                       'base_experience',
                                       "regexp_extract(species.url, r'\/(\d+)', 1) as species_id")\

df_species = df_pokemon_species_raw.selectExpr('id as species_id',
                                               'name as species_name',
                                               "cast(regexp_extract(evolution_chain.url, r'\/(\d+)', 1) as long) as evol_chain_id")

df_pokemon = df_pokemon.join(df_species, ['species_id'], 'left')\
                       .select("pokemon_id",
                               "pokemon_name",
                               "is_default",
                               "height",
                               "base_experience",
                               "evol_chain_id",
                               "species_id",
                               "species_name")

df_pokemon.display()

pokemon_id pokemon_name is_default height base_experience evol_chain_id species_id species_name 26 raichu true 8 243 10 26 raichu 29 nidoran-f true 4 55 12 29 nidoran-f 65 alakazam true 15 250 27 65 alakazam 19 rattata true 3 51 7 19 rattata 54 psyduck true 8 64 23 54 psyduck 112 rhydon true 19 170 50 112 rhydon 113 chansey true 11 395 51 113 chansey 22 fearow true 12 155 8 22 fearow 130 gyarados true 65 189 64 130 gyarados 7 squirtle true 5 63 3 7 squirtle 77 ponyta true 10 82 32 77 ponyta 34 nidoking true 14 253 13 34 nidoking 126 magmar true 13 173 61 126 magmar 50 diglett true 2 53 21 50 diglett 94 gengar true 15 250 40 94 gengar 110 weezing true 12 172 49 110 weezing 57 primeape true 10 159 24 57 primeape 136 flareon true 9 184 67 136 flareon 144 articuno true 17 290 73 144 articuno 32 nidoran-m true 5 55 13 32 nidoran-m 43 oddish true 5 64 18 43 oddish 84 doduo true 14 62 36 84 doduo 31 nidoqueen true 13 253 12 31 nidoqueen 119 seaking true 13 158 55 119 seaking 39 jigglypuff true 5 95 16 39 jigglypuff 98 krabby true 4 65 43 98 krabby 116 horsea true 4 59 54 116 horsea 25 pikachu true 4 112 10 25 pikachu 124 jynx true 14 159 59 124 jynx 95 onix true 88 77 41 95 onix 71 victreebel true 17 221 29 71 victreebel 143 snorlax true 21 189 72 143 snorlax 6 charizard true 17 267 2 6 charizard 68 machamp true 16 253 28 68 machamp 72 tentacool true 9 67 30 72 tentacool 87 dewgong true 17 166 37 87 dewgong 58 growlithe true 7 70 25 58 growlithe 107 hitmonchan true 14 159 47 107 hitmonchan 9 blastoise true 16 265 3 9 blastoise 27 sandshrew true 6 60 11 27 sandshrew 63 abra true 9 62 27 63 abra 51 dugtrio true 7 149 21 51 dugtrio 56 mankey true 5 61 24 56 mankey 52 meowth true 4 58 22 52 meowth 17 pidgeotto true 11 122 6 17 pidgeotto 79 slowpoke true 12 63 33 79 slowpoke 103 exeggutor true 20 186 45 103 exeggutor 139 omastar true 10 173 69 139 omastar 41 zubat true 8 49 17 41 zubat 28 sandslash true 10 158 11 28 sandslash 33 nidorino true 9 128 13 33 nidorino 88 grimer true 9 65 38 88 grimer 138 omanyte true 4 71 69 138 omanyte 114 tangela true 10 87 52 114 tangela 5 charmeleon true 11 142 2 5 charmeleon 1 bulbasaur true 7 64 1 1 bulbasaur 115 kangaskhan true 22 172 53 115 kangaskhan 96 drowzee true 10 66 42 96 drowzee 104 cubone true 4 64 46 104 cubone 10 caterpie true 3 39 4 10 caterpie 120 staryu true 8 68 56 120 staryu 89 muk true 12 175 38 89 muk 128 tauros true 14 172 63 128 tauros 85 dodrio true 18 165 36 85 dodrio 106 hitmonlee true 15 159 47 106 hitmonlee 111 rhyhorn true 10 69 50 111 rhyhorn 48 venonat true 10 61 20 48 venonat 67 machoke true 15 142 28 67 machoke 142 aerodactyl true 18 180 71 142 aerodactyl 100 voltorb true 5 66 44 100 voltorb 105 marowak true 10 149 46 105 marowak 131 lapras true 25 187 65 131 lapras 108 lickitung true 12 77 48 108 lickitung 44 gloom true 8 138 18 44 gloom 132 ditto true 3 101 66 132 ditto 123 scyther true 15 100 58 123 scyther 127 pinsir true 15 175 62 127 pinsir 61 poliwhirl true 10 135 26 61 poliwhirl 3 venusaur true 20 263 1 3 venusaur 37 vulpix true 6 60 15 37 vulpix 101 electrode true 12 172 44 101 electrode 83 farfetchd true 8 132 35 83 farfetchd 12 butterfree true 11 198 4 12 butterfree 141 kabutops true 13 173 70 141 kabutops 55 golduck true 17 175 23 55 golduck 74 geodude true 4 60 31 74 geodude 8 wartortle true 10 142 3 8 wartortle 62 poliwrath true 13 255 26 62 poliwrath 11 metapod true 7 72 4 11 metapod 49 venomoth true 15 158 20 49 venomoth 35 clefairy true 6 113 14 35 clefairy 80 slowbro true 16 172 33 80 slowbro 2 ivysaur true 10 142 1 2 ivysaur 66 machop true 8 61 28 66 machop 134 vaporeon true 10 184 67 134 vaporeon 76 golem true 14 223 31 76 golem 102 exeggcute true 4 65 45 102 exeggcute 4 charmander true 6 62 2 4 charmander 92 gastly true 13 62 40 92 gastly 118 goldeen true 6 64 55 118 goldeen 13 weedle true 3 39 5 13 weedle 36 clefable true 13 242 14 36 clefable 75 graveler true 10 137 31 75 graveler 78 rapidash true 17 175 32 78 rapidash 18 pidgeot true 15 216 6 18 pid

#### Pokémon - Stats
- Possui as "stats" e seus respectivos valores base, para cada Pokémon.
- Por meio da coluna `pokemon_id`, relaciona-se com o `df_pokemon`.
- *Cada "Pokémon" possui um ou mais "stats" e cada "stat" está relacionada a um ou mais "Pokémon".*

<img src="https://i.postimg.cc/GpScVMFy/pokemon-stat.png" width = "600">

In [0]:
df_pokemon_stats = df_pokemon_raw.selectExpr('id as pokemon_id',
                                             'stats')\
                                 .select('*',
                                         explode_outer('stats').alias('stats_exploded'))\
                                 .selectExpr('*',
                                             "cast(regexp_extract(stats_exploded.stat.url, r'\/(\d+)', 1) as long) as stat_id",
                                             'stats_exploded.base_stat as stat_base',
                                             'stats_exploded.stat.name as stat_name')\
                                 .distinct()\
                                 .select('pokemon_id',
                                         'stat_id',
                                         'stat_name',
                                         'stat_base')
df_pokemon_stats.display()

pokemon_id stat_id stat_name stat_base 4 1 hp 39 7 6 speed 43 22 1 hp 65 42 3 defense 70 79 6 speed 15 89 5 special-defense 100 10 2 attack 30 33 1 hp 61 74 2 attack 80 81 6 speed 45 113 4 special-attack 35 125 6 speed 105 25 6 speed 90 41 1 hp 40 65 2 attack 50 109 3 defense 95 131 2 attack 85 132 5 special-defense 48 7 2 attack 48 16 1 hp 40 17 1 hp 63 33 3 defense 57 34 5 special-defense 75 112 3 defense 120 1 2 attack 49 95 1 hp 35 17 4 special-attack 50 28 6 speed 65 55 4 special-attack 95 76 6 speed 45 107 6 speed 76 127 2 attack 125 40 4 special-attack 85 79 2 attack 65 81 3 defense 70 86 1 hp 65 111 1 hp 80 7 5 special-defense 64 34 3 defense 77 66 2 attack 80 70 5 special-defense 45 132 1 hp 48 76 3 defense 130 82 2 attack 60 84 6 speed 75 106 4 special-attack 35 107 1 hp 50 22 4 special-attack 61 30 1 hp 70 54 6 speed 55 69 6 speed 40 85 5 special-defense 60 117 1 hp 55 15 3 defense 40 57 2 attack 105 68 2 attack 130 118 5 special-defense 50 17 5 special-defense 50 68 3 defense 80 78 1 hp 65 82 1 hp 50 84 4 special-attack 35 139 5 special-defense 70 45 6 speed 50 71 5 special-defense 70 92 3 defense 30 115 2 attack 95 121 3 defense 85 128 3 defense 95 5 3 defense 58 64 1 hp 40 35 2 attack 45 35 3 defense 48 78 3 defense 70 93 4 special-attack 115 98 1 hp 30 108 3 defense 75 120 2 attack 45 47 3 defense 80 97 5 special-defense 115 137 3 defense 70 2 3 defense 63 52 4 special-attack 40 116 6 speed 60 130 6 speed 81 12 6 speed 70 56 6 speed 70 70 4 special-attack 85 111 3 defense 95 14 3 defense 50 17 6 speed 71 39 4 special-attack 45 97 3 defense 70 102 2 attack 40 123 2 attack 110 129 2 attack 10 13 6 speed 50 39 5 special-defense 25 71 4 special-attack 100 92 2 attack 35 94 2 attack 65 33 6 speed 65 43 3 defense 55 94 4 special-attack 130 129 1 hp 20 36 3 defense 73 67 1 hp 80 118 1 hp 45 142 5 special-defense 75 43 2 attack 50 63 6 speed 90 78 6 speed 105 115 6 speed 90 116 3 defense 70 143 2 attack 110 37 4 special-attack 50 80 6 speed 30 46 1 hp 35 11 4 special-attack 25 22 5 special-defense 61 98 4 special-attack 25 117 6 speed 85 35 5 special-defense 65 45 2 attack 80 88 6 speed 25 123 1 hp 70 18 2 attack 80 37 1 hp 38 38 5 special-defense 100 41 3 defense 35 50 2 attack 55 85 4 special-attack 60 91 5 special-defense 45 12 4 special-attack 90 46 2 attack 70 91 3 defense 180 98 2 attack 105 103 3 defense 85 110 5 special-defense 70 57 1 hp 65 59 1 hp 90 73 2 attack 70 141 3 defense 105 57 6 speed 95 61 6 speed 90 72 2 attack 40 135 1 hp 65 94 3 defense 60 94 6 speed 110 105 4 special-attack 50 123 3 defense 80 123 6 speed 105 129 4 special-attack 15 3 2 attack 82 20 2 attack 81 33 4 special-attack 55 66 6 speed 35 108 5 special-defense 75 1 6 speed 45 32 5 special-defense 40 44 1 hp 60 59 6 speed 95 108 6 speed 30 122 5 special-defense 120 40 1 hp 140 48 5 special-defense 55 94 5 special-defense 75 96 1 hp 60 110 6 speed 60 113 5 special-defense 105 120 3 defense 55 140 2 attack 80 3 5 special-defense 100 73 3 defense 65 142 6 speed 130 20 5 special-defense 70 112 2 attack 130 78 4 special-attack 80 81 2 attack 35 113 2 attack 5 132 2 attack 48 1 1 hp 45 32 1 hp 46 95 5 special-defense 45 136 2 attack 130 10 1 hp 45 19 1 hp 30 19 5 special-defense 35 21 5 special-defense 31 23 2 attack 60 36 2 attack 70 50 3 defense 25 53 3 defense 60 77 4 special-attack 65 106 5 special-defense 110 12 3 defense 50 41 2 attack 45 46 5 special-defense 55 60 2 attack 50 92 1 hp 30 94 1 hp 60 117 4 special-attack 95 132 4 special-attack 48 6 2 attack 84 26 2 attack 90 29 6 speed 41 53 2 attack 70 55 1 hp 80 67 6 speed 45 73 5 special-defense 120 104 1 hp 50 112 4 special-attack 45 112 6 speed 40 131 3 defense 80 16 6 speed 56 50 4 special-attack 35 56 3 defense 35 57 4 special-attack 60 84 3 defense 45 144 1 hp 90 26 4 special-attack 90 31 6 speed 76 57 3 defense 60 66 5 special-defense 35 69 4 special-attack 70 83 4 special-attack 58 92 4 special-attack 100 105 2 attack 80 15 1 hp 65 42 5 special-defense 75 50 1 hp 10 62 1 hp 90 64

#### Pokémon - Moves

- Possui os golpes ("moves") para cada Pokémon.
- Relaciona-se com o `df_pokemon` por meio da coluna `pokemon_id`.
- *Cada "Pokémon" possui um ou mais "moves" e cada `move_id` está associado a um ou mais "Pokémon"*

<img src="https://i.postimg.cc/QMKPN0L5/pokemon-move.png" width = "600">

In [0]:
df_pokemon_moves = df_pokemon_raw.selectExpr('id as pokemon_id',
                                              'moves')\
                                 .select('pokemon_id',
                                         explode_outer('moves').alias('moves'))\
                                 .selectExpr('pokemon_id',
                                             "cast(regexp_extract(moves.move.url, r'\/(\d+)', 1) as long) as move_id",
                                             'moves.move.name as move_name')\
                                 .select('pokemon_id',
                                         'move_id',
                                         'move_name')\
                                 .selectExpr('pokemon_id',
                                             'move_id',
                                             'move_name')\
                                 .distinct()

df_pokemon_moves.display()

pokemon_id move_id move_name 1 124 sludge 4 225 dragon-breath 7 59 blizzard 12 173 snore 15 42 pin-missile 2 412 energy-ball 3 414 earth-power 5 290 secret-power 8 5 mega-punch 9 200 outrage 12 405 bug-buzz 14 334 iron-defense 1 241 sunny-day 4 68 counter 6 82 dragon-rage 12 168 thief 3 237 hidden-power 3 267 nature-power 4 261 will-o-wisp 5 69 seismic-toss 7 56 hydro-pump 8 206 false-swipe 8 352 water-pulse 12 237 hidden-power 15 290 secret-power 4 214 sleep-talk 8 330 muddy-water 4 44 bite 8 237 hidden-power 8 406 dragon-pulse 9 182 protect 12 76 solar-beam 12 129 swift 3 590 confide 7 240 rain-dance 12 16 gust 4 10 scratch 4 45 growl 6 693 brutal-swing 3 189 mud-slap 4 512 acrobatics 6 497 echoed-voice 7 281 yawn 12 164 substitute 15 371 payback 5 214 sleep-talk 6 251 beat-up 6 590 confide 8 36 take-down 15 512 acrobatics 1 13 razor-wind 1 580 grassy-terrain 6 242 crunch 7 252 fake-out 9 612 power-up-punch 1 80 petal-dance 4 349 dragon-dance 7 61 bubble-beam 12 115 reflect 12 244 psych-up 15 29 headbutt 6 468 hone-claws 12 29 headbutt 1 200 outrage 3 412 energy-ball 4 251 beat-up 5 251 beat-up 6 29 headbutt 6 36 take-down 8 70 strength 9 525 dragon-tail 15 207 swagger 8 91 dig 12 78 stun-spore 1 311 weather-ball 2 216 return 3 102 mimic 4 5 mega-punch 7 55 water-gun 7 102 mimic 12 318 silver-wind 15 398 poison-jab 3 230 sweet-scent 4 445 captivate 6 206 false-swipe 7 193 foresight 9 5 mega-punch 9 117 bide 12 106 harden 1 164 substitute 2 270 helping-hand 3 206 false-swipe 5 43 leer 11 106 harden 12 171 nightmare 15 36 take-down 15 164 substitute 15 263 facade 1 115 reflect 3 70 strength 3 388 worry-seed 4 66 submission 4 315 overheat 7 401 aqua-tail 8 174 curse 9 812 flip-turn 12 33 tackle 15 64 peck 2 210 fury-cutter 2 311 weather-ball 5 216 return 9 115 reflect 9 590 confide 4 374 fling 12 117 bide 3 92 toxic 12 18 whirlwind 2 331 bullet-seed 3 526 work-up 5 82 dragon-rage 8 445 captivate 12 241 sunny-day 3 805 terrain-pulse 6 595 mystical-fire 9 203 endure 1 36 take-down 1 70 strength 3 218 frustration 3 496 round 5 280 brick-break 6 89 earthquake 9 61 bubble-beam 9 218 frustration 3 45 growl 5 66 submission 5 184 scary-face 5 468 hone-claws 7 29 headbutt 8 200 outrage 4 468 hone-claws 5 242 crunch 7 526 work-up 8 280 brick-break 9 240 rain-dance 9 396 aura-sphere 11 527 electroweb 1 263 facade 6 53 flamethrower 6 173 snore 6 535 heat-crash 7 206 false-swipe 1 497 echoed-voice 2 79 sleep-powder 2 203 endure 5 111 defense-curl 6 83 fire-spin 7 214 sleep-talk 12 93 confusion 12 366 tailwind 15 116 focus-energy 4 210 fury-cutter 5 34 body-slam 6 46 roar 6 154 fury-swipes 7 213 attract 7 352 water-pulse 9 445 captivate 1 496 round 2 526 work-up 3 74 growth 7 99 rage 12 182 protect 12 577 draining-kiss 15 76 solar-beam 1 29 headbutt 3 63 hyper-beam 3 437 leaf-storm 4 115 reflect 7 117 bide 2 282 knock-off 4 249 rock-smash 4 421 shadow-claw 6 182 protect 6 612 power-up-punch 8 812 flip-turn 9 66 submission 2 73 leech-seed 6 307 blast-burn 13 40 poison-sting 15 283 endeavor 2 38 double-edge 7 518 water-pledge 5 9 thunder-punch 6 223 dynamic-punch 6 349 dragon-dance 7 34 body-slam 7 812 flip-turn 9 36 take-down 2 115 reflect 2 803 grassy-glide 3 115 reflect 4 216 return 8 38 double-edge 8 290 secret-power 8 590 confide 12 285 skill-swap 12 483 quiver-dance 2 133 amnesia 3 148 flash 3 249 rock-smash 4 237 hidden-power 5 349 dragon-dance 7 25 mega-kick 8 55 water-gun 9 91 dig 12 240 rain-dance 2 156 rest 2 496 round 5 249 rock-smash 6 38 double-edge 9 263 facade 1 76 solar-beam 2 182 protect 5 424 fire-fang 15 102 mimic 4 108 smokescreen 6 111 defense-curl 7 93 confusion 9 70 strength 12 226 baton-pass 5 15 cut 5 213 attract 6 9 thunder-punch 8 223 dynamic-punch 9 196 icy-wind 9 249 rock-smash 9 431 rock-climb 15 92 toxic 2 202 giga-drain 4 337 dragon-claw 6 211 steel-wing 15 565 fell-stinger 2 173 snore 2 447 grass-knot 5 5 mega-punch 6 784 breaking-swipe 7 323 water-spout 1 77 poison-powder 5 70 strength 6 445 captivate 9 111 d

#### Moves - Version Groups

- Possui as características de cada move: `move_learn_method` e `version_group_id`.
- Relaciona-se com o `df_pokemon_moves` por meio da coluna `move_id` e com o `df_versions` por meio da `version_group_id`.
- *Cada "move" está relacionado a um ou mais "version_groups" e cada "version_group" está relacionado a um ou mais "moves"*.

<img src="https://i.postimg.cc/HWQKsz5x/version-group-move.png" width = "600">

In [0]:
df_moves_version_groups = df_pokemon_raw.selectExpr('moves')\
                                        .select(explode_outer('moves').alias('moves'))\
                                        .selectExpr("cast(regexp_extract(moves.move.url, r'\/(\d+)', 1) as long) as move_id",
                                                    'moves.version_group_details as version_group_details')\
                                        .select('move_id',
                                                explode_outer('version_group_details').alias('version_group_details_exploded'))\
                                        .selectExpr('move_id',
                                                    'version_group_details_exploded.move_learn_method.name as move_learn_method',
                                                    "cast(regexp_extract(version_group_details_exploded.version_group.url, r'\/(\d+)', 1) as long) as version_group_id")\
                                        .distinct()

df_moves_version_groups.display()

move_id move_learn_method version_group_id 267 egg 20 282 tutor 16 335 tutor 10 338 tutor 8 73 level-up 19 74 level-up 11 76 level-up 6 237 machine 16 320 egg 11 345 egg 5 77 level-up 12 210 machine 3 216 machine 10 474 machine 15 20 tutor 18 74 level-up 17 164 machine 16 202 egg 17 214 machine 20 216 machine 11 235 level-up 7 75 level-up 6 207 machine 4 241 machine 5 45 level-up 7 80 level-up 14 75 level-up 3 79 level-up 5 80 egg 3 115 machine 2 124 egg 14 213 machine 14 267 egg 9 447 machine 15 14 tutor 7 113 egg 5 174 egg 15 263 machine 8 438 egg 18 447 machine 18 447 machine 20 76 level-up 17 70 machine 6 104 machine 5 113 machine 20 148 machine 6 207 machine 14 263 machine 13 437 egg 14 63 machine 16 14 machine 8 15 machine 15 45 level-up 2 174 machine 4 230 machine 3 210 tutor 6 218 machine 3 219 egg 9 263 machine 18 45 level-up 5 45 level-up 15 76 level-up 2 182 machine 3 267 egg 16 338 tutor 14 523 machine 20 74 level-up 7 148 machine 8 173 tutor 9 275 egg 14 497 machine 11 33 level-up 16 34 tutor 7 73 level-up 2 73 level-up 14 148 machine 3 213 machine 11 237 machine 17 270 machine 20 290 machine 8 46 machine 14 22 level-up 19 33 level-up 12 77 level-up 1 79 level-up 16 402 level-up 20 74 level-up 8 92 machine 15 92 machine 19 104 machine 18 210 tutor 9 241 machine 14 89 machine 12 200 tutor 9 70 machine 10 237 machine 13 290 machine 5 331 machine 8 438 egg 16 46 machine 11 707 tutor 18 76 machine 13 203 machine 20 241 machine 6 63 machine 14 130 egg 18 164 machine 17 174 egg 10 237 machine 10 263 machine 9 275 egg 11 15 machine 6 188 machine 16 207 tutor 6 213 machine 20 214 machine 9 237 machine 18 76 machine 7 104 machine 13 188 machine 9 202 tutor 14 213 machine 4 216 machine 15 219 egg 7 338 tutor 16 15 machine 13 38 level-up 15 74 level-up 12 79 level-up 19 282 tutor 18 290 machine 9 77 level-up 17 156 machine 17 173 machine 20 182 machine 15 263 machine 19 63 machine 9 14 machine 20 189 tutor 9 241 machine 15 89 machine 18 14 machine 11 92 machine 17 104 machine 1 104 machine 12 156 machine 12 156 machine 18 70 machine 12 75 level-up 8 79 level-up 2 214 machine 18 218 machine 8 230 level-up 18 237 machine 8 29 machine 3 80 egg 11 173 machine 3 188 machine 6 213 machine 5 22 level-up 20 156 machine 8 218 machine 6 267 egg 18 496 machine 17 520 tutor 16 70 machine 13 73 level-up 9 235 level-up 13 237 machine 3 36 machine 2 73 level-up 18 77 level-up 13 130 egg 14 290 machine 7 345 egg 18 580 egg 15 89 machine 14 29 machine 19 75 level-up 2 189 tutor 10 207 machine 10 92 machine 1 130 egg 6 130 egg 16 173 tutor 16 388 tutor 10 38 level-up 11 204 egg 18 230 level-up 13 235 level-up 16 14 machine 15 15 machine 2 76 level-up 13 22 level-up 8 73 level-up 11 189 machine 4 204 egg 14 249 machine 8 412 machine 11 45 level-up 12 76 machine 20 164 machine 11 320 egg 5 92 machine 10 203 machine 10 249 machine 11 15 machine 9 33 level-up 20 92 machine 2 117 machine 2 216 machine 3 63 machine 8 130 egg 7 164 machine 2 202 machine 9 267 egg 8 402 tutor 10 46 machine 10 73 level-up 3 75 level-up 4 80 egg 20 113 egg 8 263 machine 17 34 machine 2 70 machine 5 580 egg 16 218 machine 5 241 machine 3 437 egg 16 174 egg 11 230 level-up 12 320 egg 16 338 tutor 10 14 machine 14 74 level-up 14 188 machine 14 202 egg 11 20 tutor 14 164 machine 15 203 egg 18 70 machine 15 164 tutor 7 174 egg 16 218 machine 13 230 level-up 4 249 machine 6 388 level-up 9 497 machine 16 80 level-up 9 77 level-up 5 80 egg 15 174 egg 9 210 tutor 10 46 machine 4 74 level-up 10 263 machine 12 275 egg 8 345 egg 8 36 level-up 19 320 egg 18 388 tutor 18 80 level-up 8 38 level-up 19 76 machine 16 113 egg 7 113 machine 17 320 egg 14 474 machine 16 76 level-up 9 89 machine 6 523 machine 18 572 level-up 17 22 level-up 12 92 machine 12 164 machine 18 237 machine 9 263 machine 7 388 level-up 15 402 machine 20 447 machine 9 89 machine 13 204 egg 10 267 machine 17 290 machine 13 520 tutor 20 580 egg 18 77 level-up 18 204 machine 20 402 tutor 16 474 machine 20 214 tutor 6 2

### Endpoint "Version groups"

#### Versions
- Possui as "versions" que compõem cada "version_group".
- Relaciona-se com a `df_moves_version_groups` por meio da coluna `version_group_id`.
- *Uma "version" está associada a apenas um "version_group", e um "version_group" possui uma ou mais "versions".*

<img src="https://i.postimg.cc/KvxWKTr5/version-group-version.png" width = "600">

In [0]:
df_versions = df_version_groups_raw.select(col('id').alias('version_group_id'),
                                           col('name').alias('version_group_name'),
                                           explode_outer('versions').alias('versions_exploded'))\
                                   .selectExpr("cast(regexp_extract(versions_exploded.url, r'\/(\d+)', 1) as long) as version_id",
                                               'versions_exploded.name as version_name',
                                               'version_group_id',
                                               'version_group_name')

df_versions.display()

version_id version_name version_group_id version_group_name 1 red 1 red-blue 2 blue 1 red-blue 3 yellow 2 yellow 4 gold 3 gold-silver 5 silver 3 gold-silver 6 crystal 4 crystal 7 ruby 5 ruby-sapphire 8 sapphire 5 ruby-sapphire 9 emerald 6 emerald 10 firered 7 firered-leafgreen 11 leafgreen 7 firered-leafgreen 12 diamond 8 diamond-pearl 13 pearl 8 diamond-pearl 14 platinum 9 platinum 15 heartgold 10 heartgold-soulsilver 16 soulsilver 10 heartgold-soulsilver 17 black 11 black-white 18 white 11 black-white 19 colosseum 12 colosseum 20 xd 13 xd 21 black-2 14 black-2-white-2 22 white-2 14 black-2-white-2 23 x 15 x-y 24 y 15 x-y 25 omega-ruby 16 omega-ruby-alpha-sapphire 26 alpha-sapphire 16 omega-ruby-alpha-sapphire 27 sun 17 sun-moon 28 moon 17 sun-moon 29 ultra-sun 18 ultra-sun-ultra-moon 30 ultra-moon 18 ultra-sun-ultra-moon 31 lets-go-pikachu 19 lets-go-pikachu-lets-go-eevee 32 lets-go-eevee 19 lets-go-pikachu-lets-go-eevee 33 sword 20 sword-shield 34 shield 20 sword-shield 35 the-isle-of-armor 21 the-isle-of-armor 36 the-crown-tundra 22 the-crown-tundra 37 brilliant-diamond 23 brilliant-diamond-and-shining-pearl 38 shining-pearl 23 brilliant-diamond-and-shining-pearl 39 legends-arceus 24 legends-arceus

#### Pokémon - Types

- Possui os "types" de cada "Pokémon".
- Relaciona-se com o `df_pokemon` por meio da coluna `pokemon_id`.
- Relaciona-se com o `df_types_damaged_types` por meio do `type_id`.
- *Um "Pokémon" pode ter um ou mais "types" e um "type" está associado a um ou mais "Pokémon"*.

<img src="https://i.postimg.cc/vTLSq8M9/pokemon-type.png" width = "600">

In [0]:
df_pokemon_types = df_pokemon_raw.selectExpr('id as pokemon_id',
                                             'types')\
                                 .select('*',
                                         explode_outer('types').alias('types_exploded'))\
                                 .selectExpr('*',
                                             'types_exploded.type.name as type_name',
                                             "cast(regexp_extract(types_exploded.type.url, r'\/(\d+)', 1) as long) as type_id")\
                                 .distinct()\
                                 .select('pokemon_id',
                                         'type_id',
                                         'type_name')

df_pokemon_types.display()

pokemon_id type_id type_name 130 3 flying 32 4 poison 103 14 psychic 123 7 bug 124 14 psychic 135 13 electric 12 7 bug 2 4 poison 112 6 rock 60 11 water 11 7 bug 46 7 bug 62 2 fighting 84 1 normal 16 3 flying 99 11 water 115 1 normal 39 1 normal 61 11 water 81 13 electric 139 11 water 58 10 fire 144 15 ice 101 13 electric 1 4 poison 22 3 flying 79 14 psychic 20 1 normal 48 7 bug 126 10 fire 70 12 grass 80 11 water 87 11 water 37 10 fire 44 12 grass 50 5 ground 93 4 poison 6 10 fire 3 12 grass 5 10 fire 78 10 fire 3 4 poison 46 12 grass 68 2 fighting 129 11 water 39 18 fairy 87 15 ice 57 2 fighting 131 11 water 18 1 normal 93 8 ghost 43 4 poison 66 2 fighting 132 1 normal 142 6 rock 49 7 bug 6 3 flying 139 6 rock 51 5 ground 15 7 bug 114 12 grass 52 1 normal 75 6 rock 55 11 water 8 11 water 29 4 poison 73 11 water 73 4 poison 64 14 psychic 122 18 fairy 33 4 poison 14 7 bug 97 14 psychic 118 11 water 41 4 poison 49 4 poison 81 9 steel 85 1 normal 122 14 psychic 76 5 ground 80 14 psychic 82 13 electric 124 15 ice 45 12 grass 91 15 ice 137 1 normal 127 7 bug 14 4 poison 62 11 water 121 11 water 27 5 ground 40 18 fairy 7 11 water 23 4 poison 90 11 water 89 4 poison 96 14 psychic 69 4 poison 105 5 ground 91 11 water 111 6 rock 54 11 water 84 3 flying 109 4 poison 113 1 normal 140 6 rock 100 13 electric 13 7 bug 17 1 normal 25 13 electric 92 4 poison 111 5 ground 143 1 normal 131 15 ice 74 5 ground 28 5 ground 79 11 water 119 11 water 134 11 water 82 9 steel 40 1 normal 140 11 water 70 4 poison 24 4 poison 72 4 poison 144 3 flying 21 1 normal 83 3 flying 102 12 grass 71 12 grass 56 2 fighting 35 18 fairy 95 5 ground 47 12 grass 123 3 flying 36 18 fairy 30 4 poison 138 11 water 141 11 water 41 3 flying 63 14 psychic 71 4 poison 9 11 water 67 2 fighting 112 5 ground 34 4 poison 94 4 poison 13 4 poison 38 10 fire 19 1 normal 83 1 normal 86 11 water 42 3 flying 108 1 normal 44 4 poison 92 8 ghost 142 3 flying 43 12 grass 65 14 psychic 141 6 rock 120 11 water 53 1 normal 4 10 fire 121 14 psychic 21 3 flying 76 6 rock 85 3 flying 95 6 rock 102 14 psychic 15 4 poison 128 1 normal 17 3 flying 16 1 normal 116 11 water 103 12 grass 42 4 poison 106 2 fighting 31 4 poison 59 10 fire 117 11 water 22 1 normal 130 11 water 69 12 grass 1 12 grass 31 5 ground 45 4 poison 72 11 water 104 5 ground 107 2 fighting 136 10 fire 12 3 flying 74 6 rock 94 8 ghost 110 4 poison 10 7 bug 34 5 ground 77 10 fire 48 4 poison 18 3 flying 125 13 electric 88 4 poison 2 12 grass 26 13 electric 47 7 bug 138 6 rock 75 5 ground 98 11 water 133 1 normal 220 5 ground 267 7 bug 242 1 normal 149 3 flying 156 10 fire 163 1 normal 229 17 dark 231 5 ground 211 4 poison 183 11 water 193 3 flying 168 4 poison 170 11 water 176 18 fairy 195 5 ground 215 17 dark 255 10 fire 169 4 poison 233 1 normal 272 11 water 207 3 flying 222 6 rock 283 7 bug 256 10 fire 151 14 psychic 214 2 fighting 284 7 bug 199 14 psychic 172 13 electric 260 5 ground 279 3 flying 186 11 water 191 12 grass 202 14 psychic 254 12 grass 179 13 electric 213 7 bug 145 13 electric 271 11 water 171 11 water 173 18 fairy 275 12 grass 252 12 grass 174 1 normal 194 5 ground 222 11 water 267 3 flying 226 3 flying 261 17 dark 166 3 flying 264 1 normal 167 7 bug 230 11 water 253 12 grass 269 4 poison 177 3 flying 247 5 ground 266 7 bug 277 3 flying 197 17 dark 167 4 poison 189 3 flying 200 8 ghost 278 3 flying 188 12 grass 189 12 grass 198 3 flying 199 11 water 196 14 psychic 153 12 grass 221 5 ground 284 3 flying 271 12 grass 249 3 flying 170 13 electric 210 18 fairy 246 5 ground 204 7 bug 257 10 fire 269 7 bug 205 7 bug 276 3 flying 235 1 normal 248 17 dark 164 1 normal 212 7 bug 201 14 psychic 250 10 fire 274 12 grass 164 3 flying 178 3 flying 203 1 normal 282 18 fairy 166 7 bug 175 18 fairy 148 16 dragon 223 11 water 285 12 grass 177 14 psychic 184 18 fairy 262 17 dark 169 3 flying 181 13 electric 227 3 flying 150 14 psychic 178 14 psychic 247 6 rock 219 10 fire 277 1 normal 188 3 flying 248 6 rock 190 1 normal 208 5 gro

### Endpoint "Types"

#### Types - Damaged Types
- Dataframe que lista, para cada tipo de "Pokémon", os "types" contra os quais ele possui vantagem ("damaged_types").
- Relaciona-se com a `df_pokemon_types` por meio da coluna `type_id`.
- *Um "type" pode ter vantagem contra nenhum ou vários "damaged_types" e um "damaged_type" está associado a um ou mais "types".*

<img src="https://i.postimg.cc/GhK5vpXn/type-type.png" width = "380">

In [0]:
df_types_damaged_types = df_types_raw.selectExpr('id as type_id',
                                              'damage_relations.double_damage_to')\
                                  .select('*',
                                          explode_outer('double_damage_to').alias('double_damage_exploded'))\
                                  .selectExpr('type_id',
                                              "cast(regexp_extract(double_damage_exploded.url, r'\/(\d+)', 1) as long) as damaged_type_id",
                                              'double_damage_exploded.name as damaged_type_name')

df_types_damaged_types.display()

type_id damaged_type_id damaged_type_name 1 null null 2 1 normal 2 6 rock 2 9 steel 2 15 ice 2 17 dark 3 2 fighting 3 7 bug 3 12 grass 4 12 grass 4 18 fairy 5 4 poison 5 6 rock 5 9 steel 5 10 fire 5 13 electric 6 3 flying 6 7 bug 6 10 fire 6 15 ice 7 12 grass 7 14 psychic 7 17 dark 8 8 ghost 8 14 psychic 9 6 rock 9 15 ice 9 18 fairy 10 7 bug 10 9 steel 10 12 grass 10 15 ice 11 5 ground 11 6 rock 11 10 fire 12 5 ground 12 6 rock 12 11 water 13 3 flying 13 11 water 14 2 fighting 14 4 poison 15 3 flying 15 5 ground 15 12 grass 15 16 dragon 16 16 dragon 17 8 ghost 17 14 psychic 18 2 fighting 18 16 dragon 18 17 dark 10001 null null 10002 null null

### Endpoint "Evolution Chains"

#### Evolution Chains - Species

- Descreve os possíveis caminhos evolutivos de cada "Pokémon Species", sendo cada caminho evolutivo formado por:
    - `evol_chain_id`: o identificador da cadeia evolutiva de cada Pokémon
    - `species_1`: espécie que o Pokémon assume na primeira forma (uma e apenas uma)
    - `species_2`: espécie que o Pokémon assume na segunda forma (0 ou várias)
    - `species_3`: espécie que o Pokémon assume na terceira forma (0 ou várias)
- Relaciona-se com o `df_pokemon` por meio da coluna `evol_chain_id` e pelas colunas de `species_id`.
- *Um Pokémon possui um e apenas um `evol_chain_id`, mas pode ter um ou mais caminhos evolutivos, logo, na tabela `df_evol_chains_species`, podemos ter mais de uma linha com o mesmo `evol_chain_id`, com caminhos diferentes para a evolução das "species".*
- *E uma "evolution_chain" está associada a pelo menos um "Pokémon".*

<img src="https://i.postimg.cc/0NF3rHc6/pokemon-evolution-chain.png" width = "600">



<img src="https://i.postimg.cc/Z5qDHhQB/pokemon-species.png" width = "600">

In [0]:
df_evol_chains_species = df_evol_chains_raw.selectExpr('id as evol_chain_id',
                                                       "chain.species as species_1",
                                                       "chain.evolves_to as species_2")\
                                           .select("evol_chain_id",
                                                   "species_1",
                                                   explode_outer(col("species_2")).alias("species_2"))\
                                           .selectExpr("evol_chain_id",
                                                       "species_1",
                                                       "species_2.species as species_2",
                                                       "species_2.evolves_to as species_3")\
                                           .select("evol_chain_id",
                                                    "species_1",
                                                    "species_2",
                                                    explode_outer(col("species_3")).alias("species_3"))\
                                           .selectExpr("evol_chain_id",
                                                       "species_1",
                                                       "species_2",
                                                       "species_3.species as species_3")\
                                           .selectExpr("evol_chain_id",
                                                      "cast(regexp_extract(species_1.url, r'\/(\d+)', 1) as long) as species_1_id",
                                                      "species_1.name as species_1_name",
                                                      "cast(regexp_extract(species_2.url, r'\/(\d+)', 1) as long) as species_2_id",
                                                      "species_2.name as species_2_name",
                                                      "cast(regexp_extract(species_3.url, r'\/(\d+)', 1) as long) as species_3_id",
                                                      "species_3.name as species_3_name")

df_evol_chains_species.display()

evol_chain_id species_1_id species_1_name species_2_id species_2_name species_3_id species_3_name 1 1 bulbasaur 2 ivysaur 3 venusaur 2 4 charmander 5 charmeleon 6 charizard 3 7 squirtle 8 wartortle 9 blastoise 4 10 caterpie 11 metapod 12 butterfree 5 13 weedle 14 kakuna 15 beedrill 6 16 pidgey 17 pidgeotto 18 pidgeot 7 19 rattata 20 raticate null null 8 21 spearow 22 fearow null null 9 23 ekans 24 arbok null null 10 172 pichu 25 pikachu 26 raichu 11 27 sandshrew 28 sandslash null null 12 29 nidoran-f 30 nidorina 31 nidoqueen 13 32 nidoran-m 33 nidorino 34 nidoking 14 173 cleffa 35 clefairy 36 clefable 15 37 vulpix 38 ninetales null null 16 174 igglybuff 39 jigglypuff 40 wigglytuff 17 41 zubat 42 golbat 169 crobat 18 43 oddish 44 gloom 45 vileplume 18 43 oddish 44 gloom 182 bellossom 19 46 paras 47 parasect null null 20 48 venonat 49 venomoth null null 21 50 diglett 51 dugtrio null null 22 52 meowth 53 persian null null 22 52 meowth 863 perrserker null null 23 54 psyduck 55 golduck null null 24 56 mankey 57 primeape null null 25 58 growlithe 59 arcanine null null 26 60 poliwag 61 poliwhirl 62 poliwrath 26 60 poliwag 61 poliwhirl 186 politoed 27 63 abra 64 kadabra 65 alakazam 28 66 machop 67 machoke 68 machamp 29 69 bellsprout 70 weepinbell 71 victreebel 30 72 tentacool 73 tentacruel null null 31 74 geodude 75 graveler 76 golem 32 77 ponyta 78 rapidash null null 33 79 slowpoke 80 slowbro null null 33 79 slowpoke 199 slowking null null 34 81 magnemite 82 magneton 462 magnezone 35 83 farfetchd 865 sirfetchd null null 36 84 doduo 85 dodrio null null 37 86 seel 87 dewgong null null 38 88 grimer 89 muk null null 39 90 shellder 91 cloyster null null 40 92 gastly 93 haunter 94 gengar 41 95 onix 208 steelix null null 42 96 drowzee 97 hypno null null 43 98 krabby 99 kingler null null 44 100 voltorb 101 electrode null null 45 102 exeggcute 103 exeggutor null null 46 104 cubone 105 marowak null null 47 236 tyrogue 106 hitmonlee null null 47 236 tyrogue 107 hitmonchan null null 47 236 tyrogue 237 hitmontop null null 48 108 lickitung 463 lickilicky null null 49 109 koffing 110 weezing null null 50 111 rhyhorn 112 rhydon 464 rhyperior 51 440 happiny 113 chansey 242 blissey 52 114 tangela 465 tangrowth null null 53 115 kangaskhan null null null null 54 116 horsea 117 seadra 230 kingdra 55 118 goldeen 119 seaking null null 56 120 staryu 121 starmie null null 57 439 mime-jr 122 mr-mime 866 mr-rime 58 123 scyther 212 scizor null null 59 238 smoochum 124 jynx null null 60 239 elekid 125 electabuzz 466 electivire 61 240 magby 126 magmar 467 magmortar 62 127 pinsir null null null null 63 128 tauros null null null null 64 129 magikarp 130 gyarados null null 65 131 lapras null null null null 66 132 ditto null null null null 67 133 eevee 134 vaporeon null null 67 133 eevee 135 jolteon null null 67 133 eevee 136 flareon null null 67 133 eevee 196 espeon null null 67 133 eevee 197 umbreon null null 67 133 eevee 470 leafeon null null 67 133 eevee 471 glaceon null null 67 133 eevee 700 sylveon null null 68 137 porygon 233 porygon2 474 porygon-z 69 138 omanyte 139 omastar null null 70 140 kabuto 141 kabutops null null 71 142 aerodactyl null null null null 72 446 munchlax 143 snorlax null null 73 144 articuno null null null null 74 145 zapdos null null null null 75 146 moltres null null null null 76 147 dratini 148 dragonair 149 dragonite 77 150 mewtwo null null null null 78 151 mew null null null null 79 152 chikorita 153 bayleef 154 meganium 80 155 cyndaquil 156 quilava 157 typhlosion 81 158 totodile 159 croconaw 160 feraligatr 82 161 sentret 162 furret null null 83 163 hoothoot 164 noctowl null null 84 165 ledyba 166 ledian null null 85 167 spinarak 168 ariados null null 86 170 chinchou 171 lanturn null null 87 175 togepi 176 togetic 468 togekiss 88 177 natu 178 xatu null null 89 179 mareep 180 flaaffy 181 ampharos 90 298 azurill 183 marill 184 azumarill 91 438 bonsly 185 sudowoodo null null 92 187 hoppip 188 skiploom 189 jumpluff 93 190 aipom 424 ambipom

<img src="https://i.postimg.cc/RZ7RFK3q/ER-diagram-pokemon.png"/>

## Perguntas

### Pergunta 1
- Quais são os maiores Pokémon? Exiba os Pokémon e a altura.

In [0]:
all = Window.partitionBy(lit(1))

df_1 = df_pokemon.select('pokemon_name',
                         'height',
                         _max('height').over(all).alias("max_height"))\
                 .where('height = max_height')\
                 .drop('max_height')

df_1.display()

pokemon_name height eternatus-eternamax 1000

### Pergunta 2

- Quantos Pokémon possuem mais de um caminho evolutivo, ou seja, mais de uma
segunda ou terceira forma? Informe a quantidade, tomando como base o número
distinto de Pokémon na sua primeira forma.

In [0]:
window_evol_chain = Window.partitionBy(df_evol_chains_species.evol_chain_id)

df_2 = df_evol_chains_species.select('evol_chain_id',
                                     count("*").over(window_evol_chain).alias("evol_chains_qty"),
                                     'species_1_id')\
                             .distinct()\
                             .join(df_pokemon, df_pokemon.species_id == df_evol_chains_species.species_1_id, 'left')\
                             .where("evol_chains_qty > 1 AND is_default = TRUE")\
                             .select("pokemon_name",
                                     "evol_chains_qty")
df_2.display()
answer = df_2.count()

print(f"Resposta: O número de Pokémon com mais de um caminho evolutivo é: {answer}.")

pokemon_name evol_chains_qty oddish 2 meowth 2 slowpoke 2 eevee 8 poliwag 2 tyrogue 3 ralts 2 wurmple 2 burmy 2 snorunt 2 nincada 2 clamperl 2 yamask 2 cosmog 2 applin 2

Resposta: O número de Pokémon com mais de um caminho evolutivo é: 15.


### Pergunta 3

- Quais são os Pokémon do tipo gelo (ice) que mais fornecem experiência ao serem
derrotados? Leve em consideração apenas os Pokémon que estão em sua forma
padrão. Exiba os Pokémon e a experiência que é fornecida.

In [0]:
df_3 = df_pokemon_types.where("type_name = 'ice'")\
                       .join(df_pokemon, ['pokemon_id'], 'left')\
                       .where("is_default = TRUE")\
                       .select("pokemon_name",
                               "base_experience",
                               _max('base_experience').over(all).alias("max_base_exp"))\
                       .where('base_experience = max_base_exp')\
                       .drop('max_base_exp')

df_3.display()

### Pergunta 4

- Encontre o golpe (move) que é mais aprendido por Pokémon pelo método level-up e em quais versões do jogo
isso acontece, levando em consideração apenas os Pokémon que estão em sua forma padrão. Em seguida, entre os
Pokémon que aprendem esse golpe (por level-up e nas versões do jogo encontrada no passo anterior), diga quais
são os Pokémon que possuem o maior valor do atributo (stat) ‘attack’. Exiba os Pokémon, as versões do jogo
em que acontece de mais Pokémon aprenderem o golpe e qual é o valor do atributo ‘attack’ deles.

In [0]:
window_move_version_group = Window.partitionBy(df_pokemon_moves.move_id, df_moves_version_groups.version_group_id)

df_4 =  df_pokemon_moves.join(df_pokemon, ['pokemon_id'], 'left')\
                        .where("is_default = TRUE")\
                        .join(df_moves_version_groups, ['move_id'], 'left').distinct()\
                        .where("move_learn_method = 'level-up'")\
                        .select('pokemon_id',
                                'pokemon_name',
                                'move_name',
                                'move_learn_method',
                                'version_group_id',
                                count("pokemon_id").over(window_move_version_group).alias('pokemon_qty'))\
                        .select('*',
                                _max('pokemon_qty').over(all).alias("max_pokemon_qty"))\
                        .where("pokemon_qty = max_pokemon_qty")\
                        .join(df_pokemon_stats, ['pokemon_id'], 'left')\
                        .where("stat_name = 'attack'")\
                        .select('*',
                                _max('stat_base').over(all).alias("max_stat_base"))\
                        .where("stat_base = max_stat_base")\
                        .join(df_versions, ['version_group_id'], 'left')\
                        .select('pokemon_name',
                                'version_name',
                                'stat_base')

df_4.display()

pokemon_name version_name stat_base kartana gold 181 kartana silver 181 kartana crystal 181 kartana ruby 181 kartana sapphire 181 kartana emerald 181 kartana firered 181 kartana leafgreen 181 kartana diamond 181 kartana pearl 181 kartana platinum 181 kartana heartgold 181 kartana soulsilver 181 kartana black 181 kartana white 181 kartana colosseum 181 kartana xd 181 kartana black-2 181 kartana white-2 181 kartana x 181 kartana y 181 kartana omega-ruby 181 kartana alpha-sapphire 181 kartana sun 181 kartana moon 181 kartana ultra-sun 181 kartana ultra-moon 181 kartana sword 181 kartana shield 181

### Pergunta 5

- Quais são as 7 evoluções que geram o maior aumento em algum dos atributos? Por
exemplo, um Pokémon com o stat ‘attack’ igual a 50 evoluir para um com o stat ‘attack’
igual a 150, o aumento seria de 100. A análise deve ser feita para todos os atributos de
todos os Pokémon, e somente aqueles na sua forma padrão. Exiba a pré-evolução, a
evolução, o atributo afetado e o quanto ele aumentou com a evolução.

In [0]:
# Considerando a observação de que para todos os Pokémons em df_pokemon
# que estão em sua forma padrão, ou seja, is_default = True, temos que
# pokemon_id = species_id. Dessa forma, podemos considerar que tanto o 
# nome quanto o id do Pokémon são iguais ao nome e ao id da sua espécie.

# Obtendo dataframe com todas as evoluções possíveis: 
# primeira forma -> segunda forma ou segunda forma -> terceira forma.
df_5 = df_evol_chains_species.selectExpr("species_1_id as pre_evolution_id",
                                         "species_1_name as pre_evolution_pokemon",
                                         "species_2_id as evolution_id",
                                         "species_2_name as evolution_pokemon")

df_5_1 = df_evol_chains_species.selectExpr("species_2_id as pre_evolution_id",
                                           "species_2_name as pre_evolution_pokemon",
                                           "species_3_id as evolution_id",
                                           "species_3_name as evolution_pokemon")
df_5 = df_5.union(df_5_1)\
           .distinct()\
           .where(col("pre_evolution_id").isNotNull()&\
                  col("evolution_id").isNotNull())\
           .join(df_pokemon_stats, df_pokemon_stats.pokemon_id == df_5.pre_evolution_id, 'left')\
           .selectExpr("pre_evolution_pokemon",
                       "stat_name as pre_stat_name",
                       "stat_base as pre_stat_base",
                       "evolution_id",
                       "evolution_pokemon")\
           .join(df_pokemon_stats, df_pokemon_stats.pokemon_id == df_5.evolution_id, 'left')\
           .selectExpr("pre_evolution_pokemon",
                       "pre_stat_name",
                       "pre_stat_base",
                       "evolution_pokemon",
                       "stat_name as stat_name_after",
                       "stat_base as stat_base_after")\
           .where("pre_stat_name = stat_name_after")\
           .selectExpr("pre_evolution_pokemon",
                       "evolution_pokemon",
                       "pre_stat_name as stat",
                       "(stat_base_after - pre_stat_base) as stat_increase")\
           .orderBy("stat_increase", ascending=False)\
           .limit(7)

df_5.display()

pre_evolution_pokemon evolution_pokemon stat stat_increase happiny chansey hp 150 nincada ninjask speed 120 shelmet accelgor speed 120 magikarp gyarados attack 115 cosmoem lunala special-attack 108 cosmoem solgaleo attack 108 cosmog cosmoem special-defense 100

### Pergunta 6

- Quais são os Pokémon com maior vantagem atacante sobre outros Pokémon? Leve em
consideração apenas o tipo do Pokémon (Um Pokémon do tipo fogo/lutador
(fire/fighting) possui vantagem atacante em um Pokémon inseto/voador (bug/flying),
pois o tipo fogo tem vantagem sobre inseto, apesar do tipo lutador não ser eficiente
contra o tipo inseto e voador) e apenas os Pokémon que estão em sua forma padrão.
Exiba os Pokémon, e a quantidade de Pokémon que eles possuem vantagem atacante
sobre.

In [0]:
df_6 = df_pokemon_types.join(df_pokemon, ['pokemon_id'], 'left')\
                       .where("is_default = TRUE")\
                       .join(df_types_damaged_types, ['type_id'], 'left')\
                       .selectExpr("pokemon_id as pokemon_id_1",
                                   "pokemon_name as pokemon_name_1",
                                   "type_name as type_pokemon",
                                   "damaged_type_id")\
                       .join(df_pokemon_types, df_pokemon_types.type_id == df_types_damaged_types.damaged_type_id, 'left')\
                       .join(df_pokemon, ['pokemon_id'], 'left')\
                       .where("is_default = TRUE")\
                       .selectExpr("pokemon_name_1 as pokemon_name",
                                   "pokemon_id as pokemon_id_damaged")\
                       .distinct()\
                       .groupBy("pokemon_name")\
                       .agg(count("pokemon_id_damaged").alias("disadvantaged_pokemon_qty"))\
                       .withColumn("max_num_pokemon", _max('disadvantaged_pokemon_qty').over(all))\
                       .where('disadvantaged_pokemon_qty = max_num_pokemon')\
                       .drop('max_num_pokemon')

df_6.display()

pokemon_name disadvantaged_pokemon_qty crabominable 546